src/shims/x86/sse41.rs

```diff
@@ -257,112 +222,34 @@ pub(super) trait EvalContextExt<'mir, 'tcx: 'mir>:
             "ptestz" | "ptestc" | "ptestnzc" => {
                 let [op, mask] = this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
 
-                let (op, op_len) = this.operand_to_simd(op)?;
-                let (mask, mask_len) = this.operand_to_simd(mask)?;
-
-                assert_eq!(op_len, mask_len);
-
-                let f = match unprefixed_name {
-                    "ptestz" => |op, mask| op & mask == 0,
-                    "ptestc" => |op, mask| op & mask == mask,
-                    "ptestnzc" => |op, mask| op & mask != 0 && op & mask != mask,
-                    _ => unreachable!(),
-                };
-
-                let mut all_zero = true;
-                for i in 0..op_len {
-                    let op = this.read_scalar(&this.project_index(&op, i)?)?.to_u64()?;
-                    let mask = this.read_scalar(&this.project_index(&mask, i)?)?.to_u64()?;
-                    all_zero &= f(op, mask);
-                }
-
-                this.write_scalar(Scalar::from_i32(all_zero.into()), dest)?;
+                let f: fn(bool, ImmTy<'_, _>, ImmTy<'_, _>) -> InterpResult<'tcx, bool> =
+                    match unprefixed_name {
+                        "ptestz" =>
+                            |acc, op, mask| {
+                                let op = op.to_scalar().to_uint(op.layout.size)?;
+                                let mask = mask.to_scalar().to_uint(mask.layout.size)?;
+                                Ok(acc && (op & mask) == 0)
+                            },
+                        "ptestc" =>
+                            |acc, op, mask| {
+                                let op = op.to_scalar().to_uint(op.layout.size)?;
+                                let mask = mask.to_scalar().to_uint(mask.layout.size)?;
+                                Ok(acc && (op & mask) == mask)
+                            },
+                        "ptestnzc" =>
+                            |acc, op, mask| {
+                                let op = op.to_scalar().to_uint(op.layout.size)?;
+                                let mask = mask.to_scalar().to_uint(mask.layout.size)?;
+                                Ok(acc && (op & mask) != 0 && (op & mask) != mask)
```

Comment:
I think the code will be overall shorter if you move the match down inside the closure; then the `let op` and `let mask` can be outside the `match`.